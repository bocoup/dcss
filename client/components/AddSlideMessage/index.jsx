import React from 'react';
import PropTypes from 'prop-types';
import { Icon, Message } from 'semantic-ui-react';

export default function AddSlideMessage({ onClick }) {
  return (
    <Message
      floating
      icon={
        <Icon.Group
          size="huge"
          style={{
            marginRight: '0.5rem !important'
          }}
        >
          <Icon name="plus square outline" />
          <Icon corner="top right" name="add" color="green" />
        </Icon.Group>
      }
      header="Click to add a slide"
      onClick={onClick}
      style={{
        cursor: 'pointer'
      }}
    />
  );
}
