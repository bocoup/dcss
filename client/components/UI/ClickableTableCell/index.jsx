import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Popup, Table } from 'semantic-ui-react';
import { detect } from 'detect-browser';

const IS_ANDROID = detect()
  .os.toLowerCase()
  .includes('android');

const ClickableTableCell = props => {
  const { children, content, href, popup, ...rest } = props;
  const onClick =
    props.onClick || (IS_ANDROID ? () => (location.href = href) : null);

  const hasChildren = children && children.length !== 0;
  const child = hasChildren ? children : content;

  const link = href && child ? <Link to={href}>{child}</Link> : child;

  const cell = (
    <Table.Cell
      {...rest}
      style={{ cursor: 'pointer' }}
      onClick={onClick}
      onTouchStart={onClick}
    >
      {link}
    </Table.Cell>
  );

  return popup ? <Popup size="tiny" content={popup} trigger={cell} /> : cell;
};

ClickableTableCell.propTypes = {
  as: PropTypes.elementType,
  active: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  collapsing: PropTypes.bool,
  content: PropTypes.any,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  icon: PropTypes.any,
  negative: PropTypes.bool,
  positive: PropTypes.bool,
  selectable: PropTypes.bool,
  singleLine: PropTypes.bool,
  textAlign: PropTypes.string,
  verticalAlign: PropTypes.string,
  warning: PropTypes.bool,
  width: PropTypes.number,
  href: PropTypes.string,
  popup: PropTypes.string,
  onClick: PropTypes.func
};

export default ClickableTableCell;
