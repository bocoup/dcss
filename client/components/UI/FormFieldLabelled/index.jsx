import React, { Component } from 'react';
import PropTypes from 'prop-types';

class FormFieldLabelled extends Component {
  render() {
    const { children, content, label = null } = this.props;
    return (
      <div className="field">
        {label && <label>{label}</label>}
        {content || children}
      </div>
    );
  }
}

FormFieldLabelled.propTypes = {
  children: PropTypes.any,
  content: PropTypes.any,
  label: PropTypes.string
};

export default FormFieldLabelled;
