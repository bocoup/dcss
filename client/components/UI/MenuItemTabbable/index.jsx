import React from 'react';
import PropTypes from 'prop-types';
import { Menu } from 'semantic-ui-react';
import Events from '@utils/Events';

const noOp = () => {};

const MenuItemTabbable = ({ children, ...props }) => {
  const onClick = props.onClick || noOp;
  const onAnyKeyUp = props.onKeyUp || noOp;
  const onKeyUp = (event, data) => {
    onAnyKeyUp(event, data);
    if (onClick !== noOp) {
      Events.onKeyUp(event, data, onClick);
    }
  };

  const tabIndex = props.tabIndex || 0;
  const menuItemProps = {
    ...props,
    onKeyUp,
    tabIndex
  };

  return (
    <Menu.Item tabIndex="0" {...menuItemProps}>
      {children}
    </Menu.Item>
  );
};

MenuItemTabbable.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  onClick: PropTypes.func,
  onKeyUp: PropTypes.func,
  tabIndex: PropTypes.any
};

export default MenuItemTabbable;
