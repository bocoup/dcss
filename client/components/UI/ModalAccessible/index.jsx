import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import FocusTrap from 'focus-trap-react';

let rootNode;

class ModalAccessible extends Component {
  componentDidMount() {
    if (!rootNode) {
      rootNode = document.getElementById('root');
    }
    rootNode.setAttribute('aria-hidden', true);
  }
  componentWillUnmount() {
    if (!rootNode) {
      rootNode = document.getElementById('root');
    }
    rootNode.setAttribute('aria-hidden', false);
  }
  render() {
    const { children, open } = this.props;
    return open ? <FocusTrap>{children}</FocusTrap> : null;
  }
}

ModalAccessible.propTypes = {
  open: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

export default ModalAccessible;
