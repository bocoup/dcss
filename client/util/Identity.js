import hash from 'object-hash';

let increment = 1;

const rando = (...props) => {
  if (props) {
    if (props.length === 1) {
      return hash(props);
    }
    return hash({ ...props });
  }

  let returnValue;
  while (
    (returnValue = Math.random()
      .toString(36)
      .slice(2))
  ) {
    if (!/^\d/.test(returnValue)) {
      return returnValue;
    }
  }
};

export function key(...props) {
  return `${rando(...props)}${increment++}`;
}

export function id() {
  return `${rando()}${increment++}`;
}

export default {
  id,
  key
};
