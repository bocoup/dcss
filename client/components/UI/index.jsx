export * from 'semantic-ui-react';

/**
 * Custom Extensions
 */
import { Table } from 'semantic-ui-react';
import ClickableTableCell from './ClickableTableCell';
Table.Cell.Clickable = ClickableTableCell;

/**
 * components
 * Supports migration to default export
 */
export * as default from 'semantic-ui-react';
