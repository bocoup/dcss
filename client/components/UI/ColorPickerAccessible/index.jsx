import React, { Fragment, useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Identity from '@utils/Identity';
import { Button, Modal, Segment } from '@components/UI';
import './ColorPickerAccessible.css';

const accessibleColorPalette = [
  '#781c81',
  '#462987',
  '#3f59a9',
  '#4686c2',
  '#58a4ac',
  '#73b580',
  '#95bd5e',
  '#b9bd4a',
  '#d6b13e',
  '#e59235',
  '#e55e2b',
  '#d92120'
];

let timeout = null;

function isMouseClick(e) {
  return !!(e && (e.screenX || e.screenY));
}

function ColorPicker(props) {
  const { disabled, name: id, name, onChange } = props;
  const [open, setOpen] = useState(false);
  const [color, setColor] = useState(props.value);
  const [focusIndex, setFocusIndex] = useState(0);
  const [shouldTrackMouse, setShouldTrackMouse] = useState(false);

  const onOpenOrCloseClick = e => {
    if (!disabled) {
      e.persist();
      // Picker opened via mouse click.
      setShouldTrackMouse(isMouseClick(e));
      setOpen(!open);
    }
  };

  const onMouseEnter = () => {
    if (!shouldTrackMouse) {
      return;
    }
    if (open) {
      clearTimeout(timeout);
      timeout = setTimeout(() => {
        setOpen(!open);
      }, 700);
    }
  };

  const onMouseLeave = () => {
    if (!shouldTrackMouse) {
      return;
    }
    clearTimeout(timeout);
  };

  const onFocus = index => {
    setFocusIndex(index);
  };

  const onClick = (e, { value }) => {
    setColor(value);
    onChange(
      {},
      {
        name,
        value
      }
    );
    onOpenOrCloseClick(e);
  };

  const ariaLabel = disabled
    ? `Select color is ${color}`
    : `Click to select a color`;

  const buttonProps = {
    'aria-label': ariaLabel,
    id,
    onClick: onOpenOrCloseClick,
    style: {
      background: color
    },
    tabIndex: 0
  };

  // TODO: move this into css
  const pickerPanelStyle = {
    background: '#ffffff',
    border: '0px solid rgba(0, 0, 0, 0.25)',
    boxShadow: 'rgba(0, 0, 0, 0.25) 0px 1px 4px',
    borderRadius: '4px',
    padding: '1em 0.5em 0.5em 1em',
    width: '315px'
  };

  return (
    <Fragment>
      <button
        className="cpa__swatch-button"
        role="button"
        {...buttonProps}
      ></button>
      {open ? (
        <div className="cpa__swatch-select-popup">
          <div
            className="cpa__swatch-select-popup-background"
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
          />
          <Modal.Accessible open={open}>
            <div style={pickerPanelStyle}>
              {accessibleColorPalette.map((background, index) => {
                const key = Identity.key({ background });
                let style = {
                  background
                };

                if (focusIndex === index) {
                  style = {
                    ...style,
                    border: '2px solid black'
                  };
                }

                const onButtonFocus = () => onFocus(index);
                return (
                  <Button
                    name="color"
                    tabIndex={0}
                    key={key}
                    value={background}
                    title={background}
                    className="cpa__swatch-select-button"
                    style={style}
                    onFocus={onButtonFocus}
                    onClick={onClick}
                  />
                );
              })}
              <div style={{ clear: 'both' }} />
            </div>
          </Modal.Accessible>
        </div>
      ) : null}
    </Fragment>
  );
}

ColorPicker.propTypes = {
  disabled: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.string
};

ColorPicker.Accessible = ColorPicker;

export { ColorPicker };
