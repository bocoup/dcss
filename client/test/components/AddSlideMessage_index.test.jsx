import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect,
}));

import assert from 'assert';
import {
  fetchImplementation,
  mounter,
  reduxer,
  snapshotter,
  state,
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import { mount, shallow } from 'enzyme';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Identity from '@utils/Identity';
jest.mock('@utils/Identity', () => {
  let count = 0;
  return {
    ...jest.requireActual('@utils/Identity'),
    id() {
      return ++count;
    },
  };
});
import AddSlideMessage from '../../components/AddSlideMessage/index.jsx';

import { Icon, Message } from '@components/UI';

const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
});

afterEach(() => {
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
});

test('AddSlideMessage', () => {
  expect(AddSlideMessage).toBeDefined();
});

test('Render 1 1', async (done) => {
  const Component = AddSlideMessage;

  const props = {
    ...commonProps,
    onClick: jest.fn(() => {}),
  };

  const state = {
    ...commonState,
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);
  const mounted = mounter(ConnectedRoutedComponent);
  expect(snapshotter(mounted)).toMatchSnapshot();
  expect(
    snapshotter(mounted.findWhere((n) => n.type() === Component))
  ).toMatchSnapshot();

  const { asFragment } = render(<ConnectedRoutedComponent {...props} />);
  expect(asFragment()).toMatchSnapshot();

  userEvent.click(screen.getByLabelText('Add a slide'));

  expect(asFragment()).toMatchSnapshot();

  expect(props.onClick.mock.calls.length).toBe(1);

  done();
});

/*{INJECTION}*/

