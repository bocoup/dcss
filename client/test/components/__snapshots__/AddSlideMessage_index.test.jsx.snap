// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
"<body>
  <div
    id=\\"root\\"
  />
  <div>
    <div
      aria-label=\\"Add a slide\\"
      class=\\"ui icon message\\"
      style=\\"cursor: pointer;\\"
    >
      <i
        aria-hidden=\\"true\\"
        class=\\"plus square outline huge icon\\"
      />
      <div
        class=\\"content\\"
      >
        <div
          class=\\"header\\"
        >
          Click to add a slide
        </div>
      </div>
    </div>
  </div>
</body>"
`;

exports[`Render 1 1 2`] = `
"<body>
  <div
    id=\\"root\\"
  />
  <div>
    <div
      aria-label=\\"Add a slide\\"
      class=\\"ui icon message\\"
      style=\\"cursor: pointer;\\"
    >
      <i
        aria-hidden=\\"true\\"
        class=\\"plus square outline huge icon\\"
      />
      <div
        class=\\"content\\"
      >
        <div
          class=\\"header\\"
        >
          Click to add a slide
        </div>
      </div>
    </div>
  </div>
</body>"
`;
